patchwork-extensions-blockentity/src/main/java/net/patchworkmc/mixin/extensions/blockentity/MixinWorldRenderer.java

```diff
@@ -0,0 +1,99 @@
+/*
+ * Minecraft Forge, Patchwork Project
+ * Copyright (c) 2016-2020, 2019-2020
+ *
+ * This library is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation version 2.1
+ * of the License.
+ *
+ * This library is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this library; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301  USA
+ */
+
+package net.patchworkmc.mixin.extensions.blockentity;
+
+import org.spongepowered.asm.mixin.Mixin;
+import org.spongepowered.asm.mixin.Shadow;
+import org.spongepowered.asm.mixin.Unique;
+import org.spongepowered.asm.mixin.injection.At;
+import org.spongepowered.asm.mixin.injection.Constant;
+import org.spongepowered.asm.mixin.injection.ModifyConstant;
+import org.spongepowered.asm.mixin.injection.Redirect;
+import net.minecraftforge.common.extensions.IForgeTileEntity;
+
+import net.minecraft.block.ChestBlock;
+import net.minecraft.block.entity.BlockEntity;
+import net.minecraft.client.render.BlockBreakingInfo;
+import net.minecraft.client.render.Camera;
+import net.minecraft.client.render.VisibleRegion;
+import net.minecraft.client.render.WorldRenderer;
+import net.minecraft.client.render.block.entity.BlockEntityRenderDispatcher;
+import net.minecraft.client.world.ClientWorld;
+import net.minecraft.util.math.BlockPos;
+
+/**
+ * Implements {@link IForgeTileEntity#getRenderBoundingBox()} and
+ * {@link IForgeTileEntity#canRenderBreaking()}.
+ */
+@Mixin(WorldRenderer.class)
+public class MixinWorldRenderer {
+	/////////////////////////////////////
+	/// renderEntities()
+	/////////////////////////////////////
+	// Redirect two calls.
+	@Redirect(method = "renderEntities", at = @At(value = "INVOKE",
+			target = "net/minecraft/client/render/block/entity/BlockEntityRenderDispatcher.render(Lnet/minecraft/block/entity/BlockEntity;FI)V"))
+	private void renderBlockEntityIfVisible(BlockEntityRenderDispatcher dispatcher, BlockEntity blockEntity, float tickDelta, int blockBreakStage,
+			Camera camera, VisibleRegion visibleRegion, float tickDeltaParam) {
+		IForgeTileEntity te = (IForgeTileEntity) blockEntity;
+
+		if (visibleRegion.intersects(te.getRenderBoundingBox())) {
+			dispatcher.render(blockEntity, tickDeltaParam, blockBreakStage);
+		}
+	}
+
+	/////////////////////////////////////
+	/// renderPartiallyBrokenBlocks()
+	/////////////////////////////////////
+	@Shadow
+	private ClientWorld world;
+	@Unique
+	private static final ThreadLocal<BlockPos> blockPosParam = ThreadLocal.withInitial(() -> null);
```

Comment:
again, i'm 90% sure this is not thread safe

Reply:
here I only collect the value of a local variable, there's no access to the internal state of other objects, I don't get why this is unsafe.